src/data/sym/basic.lean

```diff
@@ -150,4 +159,111 @@ instance inhabited_sym' [inhabited α] (n : ℕ) : inhabited (sym' α n) :=
+        rw h,
+        have ob := @wa hk rfl.ge n w begin
+          rw [h, multiset.card_cons] at d,
+          refine nat.succ.inj d,
+        end begin
+          rw multiset.card_cons at b,
+          refine nat.succ.inj b,
+        end,
+        rw [ob, g re k],
+      end } }
+end⟩
+
+instance unique {α : Type*} {n : ℕ} [g : unique α] : unique (sym α n) := unique.mk' _
+
+instance is_empty {α : Type*} {n : ℕ} [g : is_empty α] : is_empty (sym α n.succ) := ⟨begin
+  intro h,
+  rw sym at h,
+  have w := @multiset.exists_mem_of_ne_zero _ h.val begin
+    intro y,
+    have z := h.property,
+    rw [y, multiset.card_zero] at z,
+    exact false.elim (nat.succ_ne_zero n z.symm),
+  end,
+  rcases w with ⟨w, q⟩,
+  unfreezingI { cases g },
+  tauto,
+end⟩
+
+def repeat (a : α) : sym α n := ⟨multiset.repeat a n, multiset.card_repeat _ _⟩
+
+lemma repeat_inj (a b : α) : @repeat _ n.succ a = repeat b → a = b := begin
+  intro x,
+  simp only [repeat, subtype.mk.inj_eq] at x,
+  exact multiset.repeat_inj a b n x,
+end
+
+instance nontrivial {α : Type*} {n : ℕ} [g : nontrivial α] : nontrivial (sym α (n + 1)) :=
+⟨begin
+  unfreezingI { rcases g with ⟨w, ⟨m, g⟩⟩ },
+  use [@repeat _ (n + 1) w, @repeat _ (n + 1) m],
+  intro x,
+  exact g (repeat_inj _ _ x),
+end⟩
+
+def map {α β : Type*} {n : ℕ} (f : α → β) (x : sym α n) : sym β n :=
+⟨x.val.map f, by simpa [multiset.card_map] using x.property⟩
+
+@[simp] lemma mem_map {α β : Type*} {n : ℕ} {f : α → β} {b : β} {l : sym α n} :
+  b ∈ sym.map f l ↔ ∃ a, a ∈ l ∧ f a = b := multiset.mem_map
+
+@[simp] lemma map_id {α : Type*} {n : ℕ} (s : sym α n) : sym.map id s = s :=
+by simp [sym.map, subtype.mk.inj_eq]
+
+@[simp] lemma map_map {α β γ : Type*} {n : ℕ} (g : β → γ) (f : α → β) (s : sym α n) :
+  sym.map g (sym.map f s) = sym.map (g ∘ f) s :=
+by simp [sym.map, subtype.mk.inj_eq]
+
+@[simp] lemma map_zero {α β : Type*} (f : α → β) :
+  sym.map f (0 : sym α 0) = (0 : sym β 0) :=
+begin
+  rw sym.has_zero,
+  simp only [sym.map, multiset.map_zero],
+  rw sym.has_zero,
+  simp only [subtype.mk_eq_mk],
+end
+
+@[simp] lemma map_cons {α β : Type*} {n : ℕ} (f : α → β) (a : α) (s : sym α n) :
+  sym.map f (a::s) = (f a)::sym.map f s :=
+begin
+  simp only [map, subtype.mk.inj_eq, cons],
+  convert multiset.map_cons f a s.val,
+  cases s,
+  rw sym.cons,
+end
+
+lemma equiv_congr (β : Type u) (h : α ≃ β) : sym α n ≃ sym β n :=
+⟨ λ x, x.map h.to_fun,
+  λ x, x.map h.inv_fun,
+  by simp only [function.left_inverse, function.comp, map, equiv.to_fun_as_coe, multiset.map_id',
+  equiv.symm_apply_apply, implies_true_iff, eq_self_iff_true, multiset.map_map, subtype.coe_eta,
+  subtype.coe_mk, subtype.val_eq_coe, equiv.inv_fun_as_coe],
+  by simp only [function.left_inverse, function.right_inverse, function.comp, map,
+  equiv.to_fun_as_coe, multiset.map_id', implies_true_iff, eq_self_iff_true, equiv.apply_symm_apply,
+  multiset.map_map, subtype.coe_eta, subtype.coe_mk, subtype.val_eq_coe, equiv.inv_fun_as_coe] ⟩
```

Comment:
Various things to learn from this one:

* For anything much more sophisticated than an `and` / `or` / `exists` that has more than one or two fields, you should usually prefer the more readable `{ field := _ }` notation over `⟨_⟩`
* `equiv` is a type not a prop, so this should be a `def`
* The way to prove a `function.left_inverse` goal is with `intro x` or `λ x` - there's no need to unfold it with `simp`.
* You should almost never write `e.to_fun` and `e.inv_fun` - these are canonically spelt `e` and `e.symm`.

```suggestion
@[simps]
def equiv_congr {β : Type u} (e : α ≃ β) : sym α n ≃ sym β n :=
{ to_fun := map e,
  inv_fun := map e.symm,
  left_inv := λ x, by rw [map_map, equiv.symm_comp_self, map_id],
  right_inv := λ x, by rw [map_map, equiv.self_comp_symm, map_id] }
```

Reply:
I'd argue this proof is still slightly better than the one you have in the PR; β bring implicit, simps, and using rw instead of simp